crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -2330,34 +2473,84 @@ impl AuthorityPerEpochStore {
                 cache_reader,
                 commit_round,
                 previously_deferred_tx_digests,
-                last_randomness_round_written,
+                generate_randomness,
             )
             .await?;
         self.record_consensus_commit_stats(&mut batch, consensus_stats)?;
 
-        // Create a pending checkpoint if we are still accepting tx.
+        // Create pending checkpoints if we are still accepting tx.
         let should_accept_tx = if let Some(lock) = &lock {
             lock.should_accept_tx()
         } else {
+            // It is ok to just release lock here as functions called by this one are the
+            // only place that transition reconfig state, and this function itself is always
+            // executed from consensus task. At this point if the lock was not already provided
+            // above, we know we won't be transitioning state for this commit.
             self.get_reconfig_state_read_lock_guard().should_accept_tx()
         };
         let make_checkpoint = should_accept_tx || final_round;
         if make_checkpoint {
-            let pending_checkpoint = PendingCheckpoint {
+            // Filter out roots of any deferred tx.
+            for deferred in deferred_tx_roots {
+                roots.remove(&deferred);
+                randomness_roots.remove(&deferred);
+            }
+
+            // Generate pending checkpoint for regular user tx.
+            let commit_height = if self.randomness_state_enabled() {
+                commit_round * 2
+            } else {
+                commit_round
+            };
+            let pending_checkpoint = PendingCheckpointV2::V2(PendingCheckpointV2Contents {
                 roots: roots.into_iter().collect(),
                 details: PendingCheckpointInfo {
                     timestamp_ms: commit_timestamp,
-                    last_of_epoch: final_round,
-                    commit_height: commit_round,
+                    last_of_epoch: final_round && randomness_round.is_none(),
+                    commit_height,
```

Comment:
i think we should change the name of the commit_height field here since it no longer stores the commit height. Perhaps `pending_checkpoint_index` or something.

We are going to have to change this to use the subdag index for mysticeti anyway,

Reply:
Index makes it sound to me like values can't be skipped, which they can, so renamed fields/associated types to `checkpoint_height`/`CheckpointHeight`